zuul-core/src/main/java/com/netflix/zuul/netty/server/psk/TlsPskHandler.java

```diff
@@ -0,0 +1,499 @@
+
+
+        public ZuulPskServer(
+                TlsCrypto crypto,
+                Registry registry,
+                ExternalTlsPskProvider externalTlsPskProvider, ChannelHandlerContext ctx) {
+            super(crypto);
+            this.pskTimings = new PSKTimings(registry);
+            this.externalTlsPskProvider = externalTlsPskProvider;
+            this.ctx = ctx;
+        }
+
+        @Override
+        public TlsCredentials getCredentials() {
+            return null;
+        }
+
+        @Override
+        protected Vector getProtocolNames() {
+            Vector protocolNames = new Vector();
+            protocolNames.addElement(ProtocolName.HTTP_1_1);
+            protocolNames.addElement(ProtocolName.HTTP_2_TLS);
+            return protocolNames;
+        }
+
+        @Override
+        public void notifyHandshakeBeginning() throws IOException {
+            pskTimings.recordHandshakeStarting();
+            this.ctx.channel().attr(TLS_HANDSHAKE_USING_EXTERNAL_PSK).set(false);
+            // TODO: sunnys - handshake timeouts
+            super.notifyHandshakeBeginning();
+        }
+
+        @Override
+        public void notifyHandshakeComplete() throws IOException {
+            pskTimings.recordHandshakeComplete();
+            this.ctx.channel().attr(TLS_HANDSHAKE_USING_EXTERNAL_PSK).set(true);
+            super.notifyHandshakeComplete();
+            ctx.fireUserEventTriggered(SslHandshakeCompletionEvent.SUCCESS);
+        }
+
+        @Override
+        protected ProtocolVersion[] getSupportedVersions() {
+            return ProtocolVersion.TLSv13.only();
+        }
+
+        @Override
+        protected int[] getSupportedCipherSuites() {
+            return TlsUtils.getSupportedCipherSuites(
+                    getCrypto(),
+                    SUPPORTED_TLS_PSK_CIPHER_SUITE_MAP.keySet().stream()
+                            .mapToInt(Number::intValue)
+                            .toArray());
+        }
+
+        @Override
+        public ProtocolVersion getServerVersion() throws IOException {
+            return super.getServerVersion();
+        }
+
+        @Override
+        @SneakyThrows // TODO: Ask BC folks to see if getExternalPSK can throw a checked exception
+        public TlsPSKExternal getExternalPSK(Vector clientPskIdentities) {
+            byte[] clientPskIdentity = ((PskIdentity)clientPskIdentities.get(0)).getIdentity();
+            byte[] psk;
+            try{
+                this.ctx.channel().attr(CLIENT_PSK_IDENTITY_ATTRIBUTE_KEY).set(new ClientPSKIdentityInfo(clientPskIdentity));
+                psk = externalTlsPskProvider.provide(clientPskIdentity, this.context.getSecurityParametersHandshake().getClientRandom());
+            }catch (PskCreationFailureException e) {
+                throw switch (e.getTlsAlertMessage()) {
+                    case unknown_psk_identity -> new TlsFatalAlert(AlertDescription.unknown_psk_identity, "Unknown or null client PSk identity");
```

Comment:
It's unclear from the API if this fails the handshake. Can you clarify how this aborts the handshake? 
We probably need to handle things like firing `SslCloseCompletionEvent` and cleanup manually anyway?